feat(router): implement scroll position restoration

vsavkin opened the pull request.

EreckGordon commented:
@vsavkin see https://github.com/angular/angular/pull/20030/files#diff-ff3b9c3c80dbf5eb828cf6d919746075R265

`wahtItDoes` <-- typo of what may cause this note to not appear in docs?

gkalpak commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+ */
+export class RouterScrollManager implements OnDestroy {
+  private subscription: Subscription;
+  private windowObj: any = window;
```
Is this going to affect SSR?

josephliccini replied:
creeping in random angular CRs again.. sorry.. one solution to that would be a `NoopRouterScrollManager` similar to `NoopAnimationDriver`

gkalpak commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+  protected scrollIntoView(anchor: string, navigationEnd: NavigationEnd) {
+    if (this.options.anchorScrolling === 'enabled') {
+      const doc = this.dom.getDefaultDocument();
+      const el = this.dom.querySelector(doc.body, `#${anchor}`);
```
Browsers would also scroll to `a[name="${anchor}"]`, so users might expect that here too.

zimme commented:
Does this take into account, async data loading and rendering?

jasonaden commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+      private router: Router,
+      private options: {scrollPositionRestoration?: string, anchorScrolling?: string} = {}) {}
+
+  setUpScrollManager(): void {
```
Why not just have this happen in the constructor? As-is you could instantiate this class and end up running ngOnDestroy which will result in an error if `setUpScrollManager()` isn't called.

vsavkin replied:
`bootstrapListener`, where these things are set up, is pretty complex, so I prefer to explicitly invoke side effects, rather than rely on DI. That's why I prefer to keep the constructor side-effect free. Also, RouterPreloader works in the same way. I added an if statement into ngOnDestroy instead.

jasonaden replied:
+1

jasonaden commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+                  });
+                }
+
+              } else if (e instanceof NavigationEnd) {
```
Is there a possibility of this not working due to #17473? Where `NavigationEnd` can run before the UI has settled.

jasonaden commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+        scan.call(this.router.events, (s: ScrollState, e: RouterEvent) => {
+              if (e instanceof NavigationStart) {
+                const navigationSource = e.navigationSource;
+                if (navigationSource === 'popstate') {
```
I think there might be a problem with the logic here. Does the array get zeroed-out in this case (states are after `NavigationEnd`:

```
positions: [[0,0]]
source: imperative
restore: undefined
   --- scroll to 0, 10
   --- click link
positions: [[0,0], [0,10]]
source: popstate
restore: undefined
   --- back button
positions: [[0,0]]
source: popstate
restore: [0,10]
   --- forward button
positions: []
source: popstate
restore: [0,0]
```

jasonaden commented on packages/router/src/events.ts:
```diff
@@ -42,6 +46,16 @@ export class RouterEvent {
+      /** @docsNotRequired */
+      url: string,
+      /** @docsNotRequired */
+      public navigationSource: NavigationSource = 'imperative') {
```
Add to `NavigationEnd`? You're not using it there in this PR, but seems like it should be there for consistency.

vsavkin replied:
I was thinking about it. The reason why I didn't go with this is that all navigations have NavigationStart, but not all of them have NavigationEnd. So if I add the source to NavigationEnd, I'll need to add it to NavigationError and NavigationCancel. It's a bigger API change, so I decided not to do it until the need is there. Especially because the user can always get the source by using groupBy.

But if you think this is required, I'll add it.

jasonaden reviewed (changes requested):
Looks good overall. I added a couple comments to those already there.

jasonaden commented on packages/common/src/location/location.ts:
```diff
@@ -13,6 +13,7 @@ import {LocationStrategy} from './location_strategy';
 /** @experimental */
 export interface PopStateEvent {
   pop?: boolean;
+  state?: any;
```
Concrete types here would be great if it works for this use-case.

vsavkin replied:
I added types in the router package cause the router knows what is stores. The field here has to remain `any` because the location cannot depend on the router.

jasonaden replied:
👍

MRosenlind commented:
Any news on this?

jlbeard84 commented on packages/platform-server/src/domino_adapter.ts:
```diff
@@ -193,6 +193,7 @@ export class DominoAdapter extends BrowserDomAdapter {
 
   getHistory(): History { throw _notImplemented('getHistory'); }
   getLocation(): Location { throw _notImplemented('getLocation'); }
+  getWindow(): Window { throw _notImplemented('getLocation'); }
```
Looks like this might need to pass a different string into the notImplemented error.

jasonaden reviewed (approved):
Looks good. Please rebase & push. We will need approval on the public API changes in order to get this merged.

jasonaden commented:
Fixes #16911

alxhub commented:
Hello @vsavkin,

This PR fails in g3 with the following error:

```
error TS2352: Type '[Event, Event]' cannot be converted to type 'NavigationStart[]'.
  Types of property 'concat' are incompatible.
    Type '{ (...items: Event[][]): Event[]; (...items: (RouterEvent | RouteConfigLoadStart | RouteConfigLoa...' is not comparable to type '{ (...items: NavigationStart[][]): NavigationStart[]; (...items: (NavigationStart | NavigationSta...'.
      Type 'Event[]' is not comparable to type 'NavigationStart[]'.
        Type 'Event' is not comparable to type 'NavigationStart'.
          Type 'ActivationEnd' is not comparable to type 'NavigationStart'.
            Property 'navigationSource' is missing in type 'ActivationEnd'.
```

The offending code is:

```
    router.events.filter(event => event instanceof NavigationStart)
        .pairwise()
        .subscribe(navigationStartEvents => {
          const [before, after] = navigationStartEvents as NavigationStart[];
```

This seems related - let me know if you discover otherwise.

brandonroberts commented:
This will close #6595 also once it lands.

msklvsk commented:
🤔 should I wait for this to land or should I start implementing workarounds…

IgorMinar commented on packages/router/test/router_scroll_manager.spec.ts:
```diff
@@ -0,0 +1,174 @@
+
+      events.next(new NavigationStart(3, '/a', 'popstate', {navigationId: 1}));
+      events.next(new NavigationEnd(3, '/a', '/a'));
+      expect(window.scrollTo).toHaveBeenCalledWith(10, 100);
```
what will happen if we now do one of the following:

- we go forward in history "back" to '/b'? will the scroll position be reset to 20,200?
- we navigate to '/c'? will the scroll position be reset to 0,0? that's what I'd expect

IgorMinar commented on packages/router/test/router_scroll_manager.spec.ts:
```diff
@@ -0,0 +1,174 @@
+  });
+
+
+  function createScrollManager(scrollPositionRestoration: string, anchorScrolling?: string) {
```
can you change this to object literal param so that the callsites are more readable?

e.g.
```
createScrollManager({scrollPositionRestoration: 'none', anchorScrolling: 'enabled'});
```

IgorMinar commented on packages/router/test/router_scroll_manager.spec.ts:
```diff
@@ -0,0 +1,174 @@
+      events.next(new NavigationStart(1, '/a#anchor'));
+      events.next(new NavigationEnd(1, '/a#anchor', '/a#anchor'));
+
+      expect(dom.invoke).not.toHaveBeenCalled();
```
should we log a warning when the element is not found? I know that this is not how the browser does it, but if we are doing a navigation within an app we do expect that the destination exist. This expectation about the existence of the destination is much weaker in the case of a document (as opposed to application) navigation.

IgorMinar commented on packages/router/test/router_scroll_manager.spec.ts:
```diff
@@ -0,0 +1,174 @@
+         class CustomRouterScrollManager extends RouterScrollManager {
+           private navigations = new Subject<any>();
+
+           constructor(router: Router, private dataFetcher: any) {
```
This is a bad example because you would never want to couple dataFetching with scrolling this tightly.

I see, that you used this example in the api docs. Could you create a more realistic example (for both docs and this test), otherwise we'll be misleading people about how we intend them to use this api.

IgorMinar commented on packages/router/src/router_scroll_manager.ts:
````diff
@@ -0,0 +1,157 @@
+ * ```
+ * @experimental
+ */
+export class RouterScrollManager implements OnDestroy {
````
can we call this something without the "Manager" suffix? How about RouterScroller?

IgorMinar commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,157 @@
+      private router: Router,
+      private options: {scrollPositionRestoration?: string, anchorScrolling?: string} = {}) {}
+
+  setUpScrollManager(): void {
```
how about just `init()` instead of `setUpScrollManager()`?

IgorMinar reviewed (changes requested):
high level feedback:
- I don't see any mention of aux routes. I expect that aux routes would not trigger any scrolling - can you prove that via a test please?
- scrolling and scroll position are useful outside of the router use-case, so I propose that we move all the scroll handling to `@angular/common` (where we already have `Location`). This would allow aio app (angular.io) to use this code to implement scroll position restoration #17308
- the whole extension mechanism via subclassing seems too liberal and will result in too much code needed for common usecases.
  - For example yOffset feature from AngularJS should be provided out of the box because it's so common https://docs.angularjs.org/api/ng/service/$anchorScroll#yOffset
  - the deferral of scrolling is also very common use case, so we should have a way for the application code (via imperative api?) to execute scrolling that was "scheduled" by the router.

vsavkin commented:
@IgorMinar

I addressed all the comments.

* I created router-unaware ScrollService and BrowserScrollService. They can be used independently. They also support offset setting.
* I renamed RouterScollManager into RouterScroller and changed the way it can be extended. Now, it exports the scrollEvents observable which you can use to know when to scroll. I updated the docs to show how to do it in a general fashion, and in a specific case of scrolling from withing a component (which I personally think is not best idea, but it is still possible and reasy).
* I updated unit tests to be more readable.
* I updated integration tests to include the case you asked for (go back, go forward etc).

I also extracted a chunk of this PR into a separate PR (https://github.com/angular/angular/pull/21728), so ignore the first commit.

jsgoupil commented:
My comment didn't get an answer 👎

illiano commented on packages/router/src/router_scroll_manager.ts:
```diff
@@ -0,0 +1,158 @@
+              }
+            }, initState).subscribe((s: ScrollState) => {
+          if (s.restore && s.navigationEnd) {
+            this.restoreScrollPosition(s.restore, s.navigationEnd);
```
I think this would make the window jumping couple times, because of redirections int the routing configs.

CaerusKaru commented on packages/platform-server/src/domino_adapter.ts:
```diff
@@ -199,6 +199,8 @@ export class DominoAdapter extends BrowserDomAdapter {
 
   getHistory(): History { throw _notImplemented('getHistory'); }
   getLocation(): Location { throw _notImplemented('getLocation'); }
+  getWindow(): Window { throw _notImplemented('getWindow'); }
```
Can this be `getDOM().defaultView`? Domino actually supports Window objects, but they are very limited.

jasonaden commented on packages/platform-browser/src/browser/location/browser_scroll_service.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+  }
+
+  private scrollToElement(el: any): void {
```
This method scrolls both vertically and horizontally. But `offset` only takes into account vertical offset. While I think needing a left offset is going to be much less common, it seems if we're scrolling horizontally we need to support horizontal offset as well.

vsavkin replied:
Fixed it

jasonaden commented on packages/router/src/router_module.ts:
```diff
@@ -279,6 +293,32 @@ export interface ExtraOptions {
+   * * 'top'--set the scroll position to 0,0..
+   * * 'enabled'--set the scroll position to the stored position.
+   */
+  scrollPositionRestoration?: ScrollPositionRestoration;
```
Inline options for public API.

vsavkin replied:
fixed

jasonaden commented on packages/router/src/router_module.ts:
```diff
@@ -279,6 +293,32 @@ export interface ExtraOptions {
+   * * 'disabled'--does nothing (default).
+   * * 'enabled'--scrolls to the element.
+   */
+  anchorScrolling?: AnchorScrolling;
```
Inline options for public API.

vsavkin replied:
fixed

jasonaden commented on packages/router/src/router_scroller.ts:
```diff
@@ -0,0 +1,197 @@
+    this.scrollEventsSubscription = this.consumeScrollEvents();
+  }
+
+  private createScrollEvents() {
```
I feel like this could be a lot more powerful by allowing the user to pass in a stream to trigger store/scrollTo behavior. The values could default to stream of `NavigationStart` and `NavigationEnd`. Allowing this to be configurable would mean users with lots of async things kicking off that finish rendering after `NavigationEnd` don't have to create their own `RouterScroller` but could easily configure this one.

vsavkin replied:
router_scroller.spec.ts shows how to achieve that.

It's sort of reversed. You would use the constructed scroll events to trigger scrolling using the scrollService. 

It's done this way because I think constructing the scrollEvents observable is actually the tricky part. I don't think reusing the scrolling logic is that interesting, because it is is basically this:  

```
scrollEvents.subscribe(e => {
      if (e.position) {
          this.scrollService.scrollToPosition(e.position);
      } else {
          this.scrollService.scrollToAnchor(e.anchor);
      }
 })
```

I also thought about providing a function to map the scrollEvent observable, but then it felt a little bit too frameworky.  Given how easy it is to write custom scrolling logic, I decided not to do it.

IgorMinar replied:
none of this stuff is exposed as public apis as far as I can tell, the test works only because it has access to itnernal symbols. please review. I like that we keep the number of exported apis to minimum so we shouldn't change that, but maybe it would be sufficient to export the ScrollEvent as a public api if it's being passed through one of the already public router observables.

vsavkin replied:
changed it to emit scroll event via router.events (the only public observable we have) and updated the spec to reflect that

alfaproject commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * @whatItDoes Returns the current scroll position.
+   */
+  abstract currentScrollPosition(): [number, number];
```
I'd say `x, y` is pretty standard (:

vsavkin replied:
Browser APIs (e.g., scrollTo) take two positional arguments, so I think it's pretty standard, but I can change it if you want.

vsavkin commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * @whatItDoes Disables automatic scroll restoration provided by the browser.
+   */
+  abstract disableAutomaticScrollRestoration(): void;
```
Addressed it.

vsavkin commented on packages/platform-browser/src/browser.ts:
```diff
@@ -70,22 +71,16 @@ export function _document(): any {
-    Title,
+    {provide: SharedStylesHost, useExisting: DomSharedStylesHost}, DomSharedStylesHost, Testability,
+    EventManager, ELEMENT_PROBE_PROVIDERS, Meta, Title,
+    {provide: ScrollService, useClass: BrowserScrollService}
```
Fixed it

vsavkin commented on packages/platform-browser/src/browser/location/browser_scroll_service.ts:
```diff
@@ -0,0 +1,109 @@
+  /**
+   * @whatItDoes Disables automatic scroll restoration provided by the browser.
+   */
+  disableAutomaticScrollRestoration() {
```
Fixed it

vsavkin commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * @whatItDoes Configures the top offset used when scrolling to an anchor.
+   *
+   * * When given a number, the router will always use the number.
```
Fixed it

vsavkin commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * @whatItDoes Returns the current scroll position.
+   */
+  abstract currentScrollPosition(): [number, number];
```
Fixed it

IgorMinar commented on packages/platform-webworker/src/web_workers/worker/null_scroll_service.ts:
```diff
@@ -0,0 +1,39 @@
+ * @whatItDoes Provides an empty implementation of the scroll service.
+ * @experimental
+ */
+export class NullScrollService extends ScrollService {
```
wouldn't you still want to restore the scroll position when the app is running in a web worker?

that would require the scrolling to happen in the ui thread as opposed to the app thread, which is a bit tricky. right?

are you just trying to say that web workers currently don't support scrolling and scroll position restoration?

IgorMinar commented on packages/router/src/router_module.ts:
```diff
@@ -280,6 +294,32 @@ export interface ExtraOptions {
+   * Configures if the router should scroll to the element when the url has a fragment.
+   *
+   * * 'disabled'--does nothing (default).
+   * * 'enabled'--scrolls to the element.
```
we should likely make this the default in the future, right? maybe document it as such?

IgorMinar commented on packages/router/src/router_module.ts:
```diff
@@ -280,6 +294,32 @@ export interface ExtraOptions {
+   *
+   * * 'disabled'--does nothing (default).
+   * * 'top'--set the scroll position to 0,0..
+   * * 'enabled'--set the scroll position to the stored position.
```
we should likely make this the default in the future, right? maybe document it as such?

IgorMinar commented on packages/router/src/router_module.ts:
```diff
@@ -280,6 +294,32 @@ export interface ExtraOptions {
+   * Configures if the scroll position needs to be restored when navigating back.
+   *
+   * * 'disabled'--does nothing (default).
+   * * 'top'--set the scroll position to 0,0..
```
isn't scroll to top the current default?

vsavkin replied:
Currently we don't do any scroll position restoration. That's why "disabled" is the default.

IgorMinar commented on packages/router/src/router_scroller.ts:
```diff
@@ -0,0 +1,175 @@
+/**
+ * @whatItDoes Represents a scrolling event.
+ *
+ * @experimental
```
this symbol is not reexported as public api. should it be? I'm not sure, that's why I'm asking.

If we do we expect that people will *need* this type in their code for basic stuff then we should expose it. Otherwise we should understand when they would need it and decide based on that.

And if we do expose it then should it really be an interface as opposed to an abstract class so that they can use it for `instanceof` check? Take a look at the existing events and let's make the apis consistent.

vsavkin replied:
Changed it to a class, and reexported it

IgorMinar commented on packages/router/src/router_scroller.ts:
```diff
@@ -0,0 +1,175 @@
+}
+
+/**
+ * @whatItDoes Manages the scroll position by the router.
```
these docs are great but since `RouterScroller` is an implementation detail we should move these docs elsewhere. Either to the place where this is being configured or to the router guide and reference that section from the config api.

IgorMinar commented on packages/router/test/router_scroller.spec.ts:
```diff
@@ -0,0 +1,162 @@
+
+  describe('extending a scroll service', () => {
+    it('work', fakeAsync(() => {
+         const {events, scrollService, scroller} = createRouterScroller(
```
can you rename the "scroller" to "routerScroller" for better readability? thanks

vsavkin replied:
Changed the test. No longer need the router scroller here

IgorMinar reviewed (commented):
lots of smaller comments and requests for changes, but overall this looks really good.

one more thing I'd like us to research is if `window.visibleViewport` api has something we should take advantage off. Take a look at https://developers.google.com/web/updates/2017/09/visual-viewport-api especially the section where the describe the pitfalls of changing offset when users pinch to zoom on mobile devices.

It's too late to merge this in for v6, but this would make an excellent addition in 6.1.

thanks @vsavkin and @jasonaden for getting this change this far.

urish commented on tools/public_api_guard/router/router.d.ts:
```diff
@@ -123,12 +123,15 @@ export declare type Event = RouterEvent | RouteConfigLoadStart | RouteConfigLoad
     paramsInheritanceStrategy?: 'emptyOnly' | 'always';
     preloadingStrategy?: any;
+    scrollOffset?: [number, number] | (() => [number, number]);
+    scrollPositionRestoration?: 'disabled' | 'enabled' | 'top';
```
@IgorMinar I just tested this, seems like navigating to `0,0` does the "right" thing. Tested on latest Chrome, Firefox with the following code:

https://codepen.io/anon/pen/WzVOOa

Also, thank you so much for taking RTL into account when adding new features 👍

jasonaden commented on packages/platform-browser/src/browser.ts:
```diff
@@ -86,6 +87,7 @@ export function _document(): any {
     ELEMENT_PROBE_PROVIDERS,
     Meta,
     Title,
+    {provide: ScrollService, useClass: BrowserScrollService},
```
I don't think we can do this while leaving `ScrollService` in `@angular/common`. And I think it *should* be in `common`. Thoughts?

vsavkin replied:
Done

vsavkin commented on packages/platform-webworker/src/worker_app.ts:
```diff
@@ -59,18 +60,14 @@ export function setupWebWorker(): void {
     {provide: ErrorHandler, useFactory: errorHandler, deps: []},
     {provide: MessageBus, useFactory: createMessageBus, deps: [NgZone]},
     {provide: APP_INITIALIZER, useValue: setupWebWorker, multi: true},
+    {provide: ScrollService, useClass: NullScrollService}
```
Done

vsavkin commented on packages/router/src/router_module.ts:
```diff
@@ -280,6 +294,32 @@ export interface ExtraOptions {
+  anchorScrolling?: 'disabled'|'enabled';
+
+  /**
+   * Configures the top offset the router will use when scrolling to an element.
```
Fixed

vsavkin commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+   *
+   * When given a tuple with two number, the router will always use the numbers.
+   * When given a function, the router will invoke the function every time it restores scroll
+   * position.
```
Fixed it

vsavkin commented on packages/common/src/location/scroll_service.ts:
```diff
@@ -0,0 +1,42 @@
+  abstract scrollToAnchor(anchor: string): void;
+
+  /**
+   * @whatItDoes Disables automatic scroll restoration provided by the browser.
```
Fixed it

vsavkin commented:
I addressed the comments:

1. I made the ScrollService tree shakable. I talked to @jasonaden about it. The solution works, but it is not elegant. Tree shakable do not work well with provider polymorphism, when we have multiple implementations of the same provider (sort of subclassing it).
2. I renamed ScrollService into ViewportScroller.
3. I made the Router emit the Scroll event, so folks who want to implement a custom scroll restoration behavior can subscribe to router.events. It addresses the public api comments @IgorMinar made.

I haven't used window.visibleViewport because it is not supported by Safari or Edge.


CURRENTLY: I had to patch output_ast to make tests pass. @alexeagle is addressing that.

vsavkin committed:
feat(router): implement scrolling restoration service

For documentation, see `RouterModule.scrollPositionRestoration`

Fixes #13636 #10929 #7791 #6595

vsavkin commented:
@jasonaden @IgorMinar I updated the PR. Let me know if it looks good.

TobiasMalikowski commented:
Hey guys i really appreciate the work and effort you put into this! It looks like this is almost merged into the core. Can anyone pls tell me how long it will take? The page not scrolling up to top after navigating is very annoying and frustrating for users because web pages and apps are expected to behave like this.

ZenSoftware commented:
@TobiasMalikowski I have been eagerly waiting for this to be merged as well.  Igor mentioned that they were going to try to release it during the release of Angular v6.1.  So I am presuming that is fairly soon.

jasonaden commented:
[Presubmit](http://test/OCL:199362790:BASE:199362818:1528235688125:db079d4d)

IgorMinar commented:
Reviewed in person with @jasonaden. LGTM now. Thank you for all the changes. Jason will still change the commit message to add a note about how to use this feature.

jasonaden committed:
fixup! feat(router): implement scrolling restoration service

alfaproject commented on packages/common/src/viewport_scroller.ts:
```diff
@@ -0,0 +1,183 @@
+ * @whatItDoes Provides an empty implementation of the viewport scroller. This will
+ * live in @angular/common as it will be used by both platform-server and platform-webworker.
+ */
+export class NullViewportScroller implements ViewportScroller {
```
We have NoopAnimations. Should this follow the same naming? NoopViewport or so

raysuelzer commented:
I see this is closed but not merged? Is this going to be merged?

choopage commented:
How do we use this feature?

choopage commented:
https://medium.com/@sarunint/scroll-position-restoration-in-angular-5a09df8bf626

choopage commented:
https://github.com/angular/angular/issues/24547

devalmehta commented:
so is there a solution to this issue?

MikeKovetsky commented:
Thank you for this feature, @vsavkin and Angular team.
Although, most of the features work as expected, I have trouble using `scrollPositionRestoration: 'enabled'.` with query params. The scrollPosition feature scrolls viewport to the top. This causes some UX problems on filter page. 

When I switch `scrollPositionRestoration` to `disabled` it does not scroll and work as expected.
Is there a valid way to ignore queryParams change?

Splaktar commented:
@MikeKovetsky this appears to be tracked in https://github.com/angular/angular/issues/24547#issuecomment-424391897 now.
